### PR TITLE
[tune] Support RESTful API for the Web Server

### PR DESCRIPTION
**Pull request by adizim**
## What do these changes do?

Currently, Tune Web Server API is using GET commands with json payloads to specify resources/actions wanted. 

Instead, we should change the client/server API to RESTful design. This includes resource modeling, model URI's, and correct HTTP methods. 

TODO

- [x] Clean/Refactor
- [x] Add more extensive tests (optional)
- [x] Make sure CURL works
- [x] Documentation in (doc/source/tune-usage.rst)

## Related issue number
#1576

### CODE REVIEW

**Review by richardliaw — COMMENTED** [review]
Left some light comments; some of the string/uri manipulation can probably be left to some built-in urllib primitives. 

I would take a look into those to make sure you don't need to do awkward string manipulations.

File: python/ray/tune/web_server.py [hunk 1]
```diff
@@ -118,53 +188,61 @@ def trial_info(self, trial):
...
+            for trial in trial_generator.next_trials():
+                runner.add_trial(trial)
+                resource["trials"].append(self._trial_info(trial))
+            return resource
 
     return Handler
 
 
 class TuneServer(threading.Thread):
+    """A TuneServer is a thread that initializes and runs a HTTPServer. The
```
Comment (richardliaw) [inline python/ray/tune/web_server.py]:
```suggestion
    """A TuneServer is a thread that initializes and runs a HTTPServer.
    
    The
```

File: python/ray/tune/web_server.py [hunk 2]
```diff
@@ -118,53 +188,61 @@ def trial_info(self, trial):
...
+                trial_id = path.split("/")[-1]
+                try:
+                    trial = self._get_trial_by_id(trial_id)
+                except TuneError:
+                    return None
+                result = trial
+
+            return result
+
+        def _add_trials(self, name, spec):
```
[the remark this answers is not in the document]
Comment (adizim) [reply python/ray/tune/web_server.py]: Do you mean for me to add documentation on the parameters?

File: doc/source/tune-usage.rst [hunk 3]
```diff
@@ -457,29 +457,43 @@ be wrapped with ``tune.function``):
     )
 
 
-Client API
+Tune API
```
Comment (richardliaw) [inline doc/source/tune-usage.rst]: can you change this to `Tune REST API`

File: python/ray/tune/web_server.py [hunk 4]
```diff
@@ -8,103 +8,180 @@
...
 if sys.version_info[0] == 2:
+    from urlparse import urljoin, urlparse
     from SimpleHTTPServer import SimpleHTTPRequestHandler
     from SocketServer import TCPServer as HTTPServer
 elif sys.version_info[0] == 3:
+    from urllib.parse import urljoin, urlparse
     from http.server import SimpleHTTPRequestHandler, HTTPServer
 
-logger = logging.getLogger(__name__)
+LOGGER = logging.getLogger(__name__)
```
Comment (richardliaw) [inline python/ray/tune/web_server.py]: let's keep this as `logger` for consistency with rest of codebase

File: python/ray/tune/web_server.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/tune/web_server.py]: actually, it's fine

File: doc/source/tune-usage.rst [hunk 5]
```diff
@@ -457,29 +457,43 @@ be wrapped with ``tune.function``):
...
-
-.. code-block:: bash
+Tune REST API
+-------------
 
-    $ pip install requests
+You can interact with an ongoing experiment with the Tune API. The Tune API is organized around REST, which includes resource-oriented URLs, accepts form-encoded requests, returns JSON-encoded responses, and uses standard HTTP protocol.
 
-To use the Client API, you can start your experiment with ``with_server=True``:
+To allow Tune to receive and respond to your API calls, you have to start your experiment with ``with_server=True``:
```
Comment (richardliaw) [inline doc/source/tune-usage.rst]:
```suggestion
To allow an ongoing Tune experiment to receive and respond to your API calls, you have to start your experiment with ``with_server=True``:
```

File: doc/source/tune-usage.rst [hunk 6]
```diff
@@ -457,29 +457,43 @@ be wrapped with ``tune.function``):
...
 
 .. code-block:: python
 
     run_experiments({...}, with_server=True, server_port=4321)
 
-Then, on the client side, you can use the following class. The server address defaults to ``localhost:4321``. If on a cluster, you may want to forward this port (e.g. ``ssh -L <local_port>:localhost:<remote_port> <address>``) so that you can use the Client on your local machine.
+The easiest way to use the Tune API is with the built-in TuneClient. To use TuneClient, verify that you have the ``requests`` library installed:
```
Comment (richardliaw) [inline doc/source/tune-usage.rst]: yeah - to be clear, the Tune API is everything that is in `ray.tune` - we should be clear about this section being a Tune Client API (Tune REST API is also fine)

File: python/ray/tune/web_server.py [hunk 7]
```diff
@@ -8,103 +8,180 @@
...
 
         return parsed
 
 
 def RunnerHandler(runner):
     class Handler(SimpleHTTPRequestHandler):
+        """A Handler is a custom handler for TuneServer.
+
+        Handles all requests and responses coming into and from
+        the TuneServer. Built off SimpleHTTPRequestHandler which provides
```
Comment (richardliaw) [inline python/ray/tune/web_server.py]: you can remove the SimpleHTTPRequestHandler line

File: doc/source/tune-usage.rst [hunk 8]
```diff
@@ -457,29 +457,43 @@ be wrapped with ``tune.function``):
...
-----------
-
-You can modify an ongoing experiment by adding or deleting trials using the Tune Client API. To do this, verify that you have the ``requests`` library installed:
-
-.. code-block:: bash
+Tune REST API
+-------------
 
-    $ pip install requests
+You can interact with an ongoing experiment with the Tune API. The Tune API is organized around REST, which includes resource-oriented URLs, accepts form-encoded requests, returns JSON-encoded responses, and uses standard HTTP protocol.
```
Comment (richardliaw) [inline doc/source/tune-usage.rst]: switch Tune API with "Tune Client API"

File: doc/source/tune-usage.rst [hunk 9]
```diff
@@ -457,29 +457,43 @@ be wrapped with ``tune.function``):
     )
 
 
-Client API
-----------
-
-You can modify an ongoing experiment by adding or deleting trials using the Tune Client API. To do this, verify that you have the ``requests`` library installed:
-
-.. code-block:: bash
+Tune REST API
```
Comment (richardliaw) [inline doc/source/tune-usage.rst]:
```suggestion
Tune Client API
```

sorry for the confusion

File: python/ray/tune/web_server.py [hunk 10]
```diff
@@ -118,62 +195,55 @@ def trial_info(self, trial):
...
     DEFAULT_PORT = 4321
 
     def __init__(self, runner, port=None):
-
+        """Initialize HTTPServer and serve forever by invoking self.run()"""
         threading.Thread.__init__(self)
         self._port = port if port else self.DEFAULT_PORT
         address = ('localhost', self._port)
-        logger.info("Starting Tune Server...")
+        LOGGER.info("Starting Tune Server...")
```
Comment (richardliaw) [inline python/ray/tune/web_server.py]: also include `self.daemon = True`

File: python/ray/tune/web_server.py [hunk 11]
```diff
@@ -8,103 +8,180 @@
 import threading
 
 import ray.cloudpickle as cloudpickle
-from ray.tune.error import TuneError, TuneManagerError
+from ray.tune import TuneError
```
Comment (richardliaw) [inline python/ray/tune/web_server.py]: can you also get rid of TuneManagerError in tune/error.py?

File: doc/source/tune-usage.rst (hunk 6 above)
Comment (richardliaw) [reply doc/source/tune-usage.rst]: @adizim

**Review by richardliaw — APPROVED** [review]

**Pull request merged**